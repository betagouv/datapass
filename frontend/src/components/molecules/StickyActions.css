.datapass-sticky-actions {
  position: fixed;
  bottom: 0;
  right: 2.5rem;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  z-index: 1;
}

.datapass-sticky-actions .datapass-sticky-actions-buttons .fr-btn {
  border: 1px solid var(--text-action-high-blue-france);
  box-shadow: 0 20px 25px -5px rgb(0 0 0 / 0.1),
    0 8px 10px -6px rgb(0 0 0 / 0.1);
}

.datapass-sticky-actions .datapass-sticky-actions-buttons .fr-btn:hover {
  background-color: var(--text-action-high-blue-france) !important;
}

.datapass-sticky-actions
  .datapass-sticky-actions-buttons
  .fr-btn--quaternary:hover {
  background-color: var(--info-950-100-hover) !important;
}

.datapass-sticky-actions-dialog {
  width: 300px;
  box-shadow: 0 20px 25px -5px rgb(0 0 0 / 0.1),
    0 8px 10px -6px rgb(0 0 0 / 0.1);
  height: 60vh;
  background-color: white;
  border: 2px solid var(--text-action-high-blue-france);
  display: flex;
  flex-direction: column;
  margin-bottom: 0.5rem;
}

.datapass-sticky-actions-dialog-header {
  display: flex;
  background-color: var(--text-action-high-blue-france);
  justify-content: space-between;
  align-items: center;
  color: white;
  padding: 0 0.5rem;
  font-size: small;
}

.datapass-sticky-actions-dialog-header-action {
  color: white;
  font-size: small;
  background-color: transparent;
  text-decoration: underline;
}

.datapass-sticky-actions-dialog-header-action:hover {
  background-color: transparent !important;
}

.datapass-sticky-actions-dialog-body {
  flex-grow: 1;
  overflow: scroll;
  display: flex;
  flex-direction: column-reverse;
  padding: 1rem 0;
}

.datapass-instruct-dialog {
  display: flex;
  flex-direction: column;
  padding: 0.5rem;
  gap: 0.5rem;
}

.datapass-instruct-dialog > * {
  width: 100%;
  justify-content: center;
}

.datapass-notify-dialog {
  padding: 0.5rem 1rem;
}

.datapass-notify-dialog textarea {
  height: 1.75rem !important;
  min-width: 100%;
  max-width: 100%;
}

.datapass-notify-dialog .fr-input-group {
  margin-bottom: 0.5rem;
}

.datapass-notify-dialog .fr-btn {
  margin-bottom: 0;
}

.datapass-notify-dialog-messages {
  margin-bottom: 1rem;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: 0.5rem;
}

.datapass-notify-dialog-message {
  display: flex;
  flex-direction: column;
  width: 70%;
}

.datapass-notify-dialog-message-from-user {
  margin-left: auto;
}

.datapass-notify-dialog-message-from-user
  .datapass-notify-dialog-message-content {
  background-color: var(--blue-cumulus-950-100);
}

.datapass-notify-dialog-message-content {
  background-color: var(--purple-glycine-975-75);
  padding: 0.5rem;
}

.datapass-notify-dialog-message-footer {
  font-size: small;
  font-style: italic;
  color: var(--text-mention-grey);
  margin-left: auto;
}
