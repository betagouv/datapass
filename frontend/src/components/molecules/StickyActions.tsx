import { FunctionComponent, useState } from 'react';
import {
  EnrollmentEvent,
  PromptType,
  eventConfigurations,
} from '../../config/event-configuration';
import EventButton from '../atoms/EventButton';
import ButtonGroup from './ButtonGroup';
import './StickyActions.css';
import Button from '../atoms/hyperTexts/Button';
import { useFormSubmission } from '../templates/Form/SubmissionPanel/hooks/use-form-submission';
import { processEvent } from '../../lib/process-event';
import Loader from '../atoms/Loader';
import Prompt from '../templates/Form/SubmissionPanel/Prompt';
import ConfirmationModal from '../organisms/ConfirmationModal';
import { chain, isEmpty } from 'lodash';
import { getChangelog } from '../../lib';
import moment from 'moment';
import { useAuth } from '../organisms/AuthContext';

export const listAuthorizedEvents = (acl: Record<string, boolean>) =>
  (Object.keys(eventConfigurations) as EnrollmentEvent[]).filter(
    (key) => acl[key]
  );

type StickyActionsProps = {
  enrollment: any;
  handlePostEvent: Function;
  updateEnrollment: Function;
};

type StickyActionsDialogProps = {
  title: string;
  onClose: Function;
  body: any;
};

type EnrollmentAction =
  | EnrollmentEvent.instruct
  | EnrollmentEvent.notify
  | null;

const StickyActionsDialog: FunctionComponent<StickyActionsDialogProps> = ({
  title,
  body,
  onClose,
}) => {
  return (
    <div className="datapass-sticky-actions-dialog">
      <div className="datapass-sticky-actions-dialog-header">
        <div className="datapass-sticky-actions-dialog-header-title">
          {title}
        </div>
        <Button
          className="datapass-sticky-actions-dialog-header-action"
          onClick={() => onClose()}
          tertiaryNoOutline
        >
          Réduire
        </Button>
      </div>
      <div className="datapass-sticky-actions-dialog-body">{body}</div>
    </div>
  );
};

export const StickyActions: FunctionComponent<StickyActionsProps> = ({
  enrollment,
  updateEnrollment,
  handlePostEvent,
}) => {
  const {
    loading,
    pendingEvent,
    onEventButtonClick,
    onPromptConfirmation,
    onPromptCancellation,
  } = useFormSubmission(
    handlePostEvent,
    enrollment,
    updateEnrollment,
    processEvent
  );

  const { user } = useAuth();
  const [currentAction, setCurrentAction] = useState<EnrollmentAction>(null);
  const authorizedEvents = listAuthorizedEvents(enrollment.acl);

  const handleActionChange = (action: EnrollmentAction) => {
    if (currentAction !== action) {
      setCurrentAction(action);
    } else {
      setCurrentAction(null);
    }
    onPromptCancellation();
  };

  const getContent = (currentAction: EnrollmentAction) => {
    switch (currentAction) {
      case EnrollmentEvent.instruct:
        return {
          title: 'Instruire',
          body: (
            <div className="datapass-instruct-dialog">
              <div>Comment souhaitez vous instruire cette demande ?</div>
              {authorizedEvents
                .filter((event) => event !== EnrollmentEvent.notify)
                .map((event) => {
                  const eventConfiguration = eventConfigurations[event];
                  return (
                    <EventButton
                      key={event}
                      disabled={!!pendingEvent || loading}
                      onClick={() => onEventButtonClick(event)}
                      {...eventConfiguration.displayProps}
                    />
                  );
                })}

              {loading && <Loader enableBePatientMessage />}

              {pendingEvent &&
                eventConfigurations[pendingEvent].prompt ===
                  PromptType.comment && (
                  <Prompt
                    onAccept={onPromptConfirmation}
                    onCancel={onPromptCancellation}
                    displayProps={
                      eventConfigurations[pendingEvent!].displayProps
                    }
                    selectedEvent={pendingEvent as string}
                    enrollment={enrollment}
                  />
                )}
              {pendingEvent &&
                eventConfigurations[pendingEvent].prompt ===
                  PromptType.submit_instead && (
                  <ConfirmationModal
                    title="Vos modifications vont être enregistrées."
                    confirmLabel="Soumettre la demande"
                    handleConfirm={onPromptConfirmation}
                    cancelLabel="Non, je l’enregistre seulement"
                    handleCancel={onPromptCancellation}
                  >
                    <p>
                      Souhaitez-vous soumettre votre habilitation à validation,
                      afin qu'elle soit étudiée par les équipes compétentes ?
                    </p>
                  </ConfirmationModal>
                )}
              {pendingEvent &&
                eventConfigurations[pendingEvent].prompt ===
                  PromptType.confirm_deletion && (
                  <ConfirmationModal
                    title="La suppression d'une habilitation est irréversible"
                    handleCancel={onPromptCancellation}
                    handleConfirm={onPromptConfirmation}
                  >
                    Voulez vous continuer ?
                  </ConfirmationModal>
                )}
              {pendingEvent &&
                eventConfigurations[pendingEvent].prompt ===
                  PromptType.confirm_archive && (
                  <ConfirmationModal
                    title="Vous n’aurez plus accès à cette habilitation"
                    handleCancel={onPromptCancellation}
                    handleConfirm={onPromptConfirmation}
                  >
                    Voulez vous continuer ?
                  </ConfirmationModal>
                )}
            </div>
          ),
        };

      case EnrollmentEvent.notify:
        let messages = chain(enrollment.events)
          .sortBy('created_at')
          .reject(
            ({ name, diff }) => name !== 'notify' && isEmpty(getChangelog(diff))
          )
          .value();

        return {
          title: 'Écrire au demandeur',
          body: (
            <div className="datapass-notify-dialog">
              <div className="datapass-notify-dialog-messages">
                {messages.map((message) => {
                  const isFromUser = message.user.id === user?.id;
                  return (
                    <div
                      key={message.id}
                      className={`datapass-notify-dialog-message ${
                        isFromUser && 'datapass-notify-dialog-message-from-user'
                      }`}
                    >
                      <div className="datapass-notify-dialog-message-content">
                        {message.comment}
                      </div>
                      <div className="datapass-notify-dialog-message-footer">
                        {moment(message.created_at).calendar()}
                      </div>
                    </div>
                  );
                })}
              </div>
              {pendingEvent &&
                eventConfigurations[pendingEvent].prompt ===
                  PromptType.notify && (
                  <Prompt
                    alignButtons="right"
                    onAccept={onPromptConfirmation}
                    displayProps={
                      eventConfigurations[pendingEvent!].displayProps
                    }
                    selectedEvent={pendingEvent as string}
                    enrollment={enrollment}
                  />
                )}
            </div>
          ),
        };

      default:
        return null;
    }
  };

  const content = getContent(currentAction);

  return (
    <div className="datapass-sticky-actions">
      {currentAction && content && (
        <StickyActionsDialog
          title={content.title}
          body={content.body}
          onClose={() => {
            handleActionChange(null);
          }}
        />
      )}
      <ButtonGroup className="datapass-sticky-actions-buttons" align="right">
        {user && user.roles.length > 1 && authorizedEvents.length > 1 && (
          <EventButton
            onClick={() => handleActionChange(EnrollmentEvent.instruct)}
            label="Instruction"
            icon="edit"
            quaternary={currentAction !== EnrollmentEvent.instruct}
            iconFill
          />
        )}
        {authorizedEvents.includes(EnrollmentEvent.notify) && (
          <EventButton
            onClick={() => {
              handleActionChange(EnrollmentEvent.notify);
              onEventButtonClick(EnrollmentEvent.notify);
            }}
            label="Messagerie"
            icon="mail"
            quaternary={currentAction !== EnrollmentEvent.notify}
            iconFill
          />
        )}
      </ButtonGroup>
    </div>
  );
};

export default StickyActions;
