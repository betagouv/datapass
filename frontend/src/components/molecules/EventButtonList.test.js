import { listAuthorizedEvents } from './StickyActions';

describe('StickyActions', () => {
  it('lists button to display from enrollment acl', () => {
    const acl = {
      create: false,
      destroy: false,
      index: false,
      notify: true,
      refuse: true,
      request_changes: false,
      submit: true,
      archive: true,
      show: true,
      update: true,
      validate: false,
    };

    expect(listAuthorizedEvents(acl)).toMatchSnapshot();
  });
});
