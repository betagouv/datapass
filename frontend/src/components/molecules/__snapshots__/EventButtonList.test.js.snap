// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventButtonList lists button to display from enrollment acl 1`] = `
Array [
  "notify",
  "update",
  "submit",
  "refuse",
  "archive",
]
`;
